migtools/oadp-non-admin
# Delete Velero Backup when user sets the deleteVeleroBackup

Enhancement which will delete original VeleroBackup when the user sets the deleteVeleroBackup witin NonAdminBackup Spec.

Fixes #58

This allows NonAdminBackup object to have additional field within spec which will trigger deletion of the original VeleroBackup object:

```yaml
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: example
  namespace: nacproject
spec:
  deleteVeleroBackup: true # <<<<<< NEW 
  [...]
```

## To test:
-------
Test 1:
deleteVeleroBackup on non existing Velero Backup object
-------
1. Create NonAdminBackup without properspec

```yaml
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: backupwithoutspec
  namespace: nacproject
spec: {}
```

2. Ensure NonAdminBackup is in:
  - Phase:
    BackingOff
  - Has the Condition:
      Type: Accepted
      Status: False
      Reason: InvalidBackupSpec
      Message: BackupSpec is not defined

3. Update the NonAdminBackup object to include:
```yaml
[...]
spec:
  deleteVeleroBackup: true
[...]
```

4. Expected result:
   - NonAdminBackup object should be deleted by the controller.
   - Reconcile loop should exit with the following DEBUG log messages:
     DEBUG  Deleted NonAdminBackup: backupwithoutspec
     DEBUG  NonAdminBackupPredicate: Accepted Delete event
     DEBUG  >>> Reconcile NonAdminBackup - loop start
     DEBUG  Non existing NonAdminBackup CR

-------
Test 2:
deleteVeleroBackup on existing Velero Backup object set to true
-------
1. Create NonAdminBackup without properspec

```yaml
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: backupwithspec
  namespace: nacproject
spec:
  backupSpec: {}
```

2. Ensure NonAdminBackup is in:
  - Phase:
    Created
  - Has the Conditions:
    - Type: Accepted
      Status: True
      Reason: BackupAccepted
      Message: Backup accepted
    - Type: Accepted
      Status: True
      Reason: BackupAccepted
      Message: Backup accepted
      
3. Ensure VeleroBackup is created in the openshift-adp namespace:
   $ oc get Backup -n openshift-adp nab-nacproject-eeaa9cd35c77f4
   NAME                            AGE
   nab-nacproject-eeaa9cd35c77f4   12s

4. Update the NonAdminBackup object to include:
```yaml
[...]
spec:
  deleteVeleroBackup: true
[...]
```

5. Expected result:
   - VeleroBackup should be deleted by the controller
   - NonAdminBackup object should be deleted by the controller.
   - Reconcile loop should exit with the following DEBUG log messages:
     DEBUG  Deleted NonAdminBackup: backupwithoutspec
     DEBUG  NonAdminBackupPredicate: Accepted Delete event
     DEBUG  >>> Reconcile NonAdminBackup - loop start
     DEBUG  Non existing NonAdminBackup CR


-------
Test 3:
deleteVeleroBackup on existing Velero Backup object set to false
-------
1. Perform the same steps as 1,2,3 from the Test 2
2. Update the NonAdminBackup object to include:
```yaml
[...]
spec:
  deleteVeleroBackup: false
[...]
```
3. Reconcile loop should have the DEBUG logs:
   DEBUG: NonAdminBackup BackupSpec and BackupStatus - nothing to update


-------
Test 4:
deleteVeleroBackup on NON existing Velero Backup object set to true
-------
1. Perform the same steps as 1,2,3 from the Test 2
2. Delete Velero Backup object:
  $ oc delete -n openshift-adp nab-nacproject-eeaa9cd35c77f4
3. Update the NonAdminBackup object to include:
```yaml
[...]
spec:
  deleteVeleroBackup: true
[...]
```
4. Expected result:
   - NonAdminBackup object should be deleted by the controller.
   - Reconcile loop should exit with the following DEBUG log messages:
     DEBUG  Deleted NonAdminBackup: backupwithoutspec
     DEBUG  NonAdminBackupPredicate: Accepted Delete event
     DEBUG  >>> Reconcile NonAdminBackup - loop start
     DEBUG  Non existing NonAdminBackup CR
